src/lib/operators/difference.cpp

```diff
@@ -35,6 +35,8 @@ std::shared_ptr<const Table> Difference::_on_execute() {
 
   auto output = std::make_shared<Table>(input_table_left()->column_definitions(), TableType::References);
 
+  if (input_table_left()->is_validated() && input_table_right()->is_validated()) output->mark_as_validated();
```

Comment:
why do you have to do this in every individual operator? cannot AbstractOperator::execute() do this?

Reply:
That's what I thought in between to, but the table is const once it leaves the operator.

Reply:
I don't see how an `std::shared_ptr<const SomeThing>` is different from a `const SomeThing`. Unless you are suggesting to remove const from everything, I don't see why we should make a difference there.

Generally, I like the idea of "only the operator itself knows how to construct the result, so no one else should be allowed to touch it".

Reply:
(But yes, it would be nicer for the AbstractOperator to handle that. It is just that given the option of saving those ~20 lines or weakening the const safety, I tend to value the latter more)

Reply:
Ignore my rant about `std::shared_ptr<const SomeThing>` that's a different topic. 

I aggree that only the operators should touch the tables. But `AbstractOperator::execute()` IS still pretty close to being the operator, so i wouldn't see a problem with `AbstractOperator::_on_execute()` handing out a mutable table that  `AbstractOperator::execute()` can mark as validated.

Doing it as it is done right now is easily forgotten when implementing an operator, I'm sure you'll agree. If validated/not_validated is an integral part of the table state, then maybe it should be a constructor argument, so setting it cannot be missed?

Reply:
>  I tend to value the latter more

Where is the value the Table being const as soon as it is obtained by AbstractOperator::execute()? I think the risk that AbstractOperator::execute() is going to do something invalid with it is very low.

Forgetting to call mark_as_validated in an operator is a very real risk, on the other hand, imho.

Reply:
I'll give this some thought in a quite minute.

Reply:
We have never used Print's in the middle of plans and I wouldn't be able to see use for it, anyway. Speaking of which, we never use Print as an operator, do we? 

I wouldn't allow Print to be the issue here. As a band-aid, its on_execute should just return a nullptr

Reply:
It's not automatically placed into a plan, but is a useful debug feature. I have placed Print into the middle of query plans quite often to have a look at the data that is passed around.

Reply:
I am pretty sure I am going to be relentless here: Requiring operators to **not forget** to "forward" a property that has nothing, really nothing, to do with the logic they implement is a bad idea. If our architecture forces us to make this requirement, then it is our architecture is bad.

(In this case, I think, the bad architecture is either: Operators, glorified functions, after all, return an immutable object. OR: the "is_validated" property on a Table is not a good idea in the first place.)

Reply:
The bug mentioned in the issue could have been avoided by Assert()ing that the Table with Insert-data has no MVCCData, right? Maybe just doing that would be the better approach to solve the issue? Tbh, I think the motivation in the issue is too weak to warrant hacky changes in all operators.

Reply:
> The bug mentioned in the issue could have been avoided by Assert()ing that the Table with Insert-data has no MVCCData, right?

No, this is all about Delete and has nothing to do with the insert.

> Tbh, I think the motivation in the issue is too weak to warrant hacky changes in all operators.

Yes. And your suggestion above led me to think about this again. Simple solution coming up in a couple of minutes. Can't believe I didn't think of that before.